File: buildSrc/src/main/java/wolkenschloss/task/StatusBuilder.java
package wolkenschloss.task;

import java.util.function.Function;
import java.util.function.Predicate;

class StatusBuilder<T> implements Check<T>, OkMessage<T>, ErrorMessage<T>, StatusChecker<T> {

    private final StatusTask statusTask;
    private final CheckedSupplier<T> fn;
    private Predicate<T> predicate;
    private String errorMessage;
    private Function<T, String> okMessageProducer;

    public StatusBuilder(StatusTask statusTask, CheckedSupplier<T> fn) {
        this.statusTask = statusTask;
        this.fn = fn;
    }

    public OkMessage<T> check(Predicate<T> p) {
        this.predicate = p;
        return this;
    }

    public ErrorMessage<T> ok(Function<T, String> message) {
        this.okMessageProducer = message;
        return this;
    }

    public ErrorMessage<T> ok(String message) {
        this.okMessageProducer = (Void) -> message;
        return this;
    }

    public StatusChecker<T> error(String message) {
        this.errorMessage = message;
        return this;
    }

    public void run(String label) {
        T value = null;
        try {
            value = fn.apply();
        } catch (Throwable throwable) {
            statusTask.getLogger().error(String.format("✗ %-15s: %s", label, throwable.getMessage()));
            return;
        }

        if (predicate.test(value)) {
            statusTask.getLogger().quiet(String.format("✓ %-15s: %s", label, okMessageProducer.apply(value)));
        } else {
            statusTask.getLogger().error(String.format("✗ %-15s: %s", label, errorMessage));
        }
    }
}

File: buildSrc/src/main/kotlin/gradle/testbed/status/StatusBuilder.kt
package family.haschka.wolkenschloss.gradle.testbed.status

import org.gradle.api.GradleException
import java.util.*
import java.util.function.Function
import java.util.function.Predicate

internal class StatusBuilder<T>(statusTask: Status, fn: () -> T) :
    Check<T>, OkMessage<T>, ErrorMessage<T>, StatusChecker {
    private val statusTask: Status
    private val fn: () -> T
    private var predicate: Optional<Predicate<T>> = Optional.empty()
    private var errorMessage: Optional<String> = Optional.empty()
    private var okMessageProducer: Optional<Function<T, String>> = Optional.empty()

    init {
        this.statusTask = statusTask
        this.fn = fn
    }

    override fun check(p: Predicate<T>): OkMessage<T> {
        predicate = Optional.of(p)
        return this
    }

    override fun ok(message: Function<T, String>): ErrorMessage<T> {
        okMessageProducer = Optional.of(message)
        return this
    }


    override fun error(message: String): StatusChecker {
        errorMessage = Optional.of(message)
        return this
    }

    override fun run(label: String) {
        val value: T  = try {
            fn()
        } catch (throwable: Throwable) {
            statusTask.logger.error(String.format("✗ %-15s: %s", label, throwable.message))
            return
        }

        predicate.filter { it.test(value) }.ifPresentOrElse({
            okMessageProducer.map { it.apply(value) }
                .map { String.format("✓ %-15s: %s", label, it) }
                .ifPresentOrElse( {
                    statusTask.logger.quiet(it)
                },
                    {
                        errorMessage.ifPresent {
                            statusTask.logger.error(String.format("✗ %-15s: %s", label, it))
                        }
                    })}, {
                        throw GradleException("No test for $label")
        })
    }
}